File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/playground/PlaygroundScaffold.kt
package xyz.ksharma.prisma.catalogue.playground

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.components.button.PrismaButton
import xyz.ksharma.prisma.components.icons.PrismaIcons
import xyz.ksharma.prisma.coreui.themed
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

/**
 * Storybook-style scaffold shared by every legacy showcase. Three pills sit
 * directly under the preview — Edit / A11y / Code — each opening its own
 * bottom sheet. State galleries (legacy) and structured A11yReports (new)
 * are both supported so showcases can adopt the new pattern progressively.
 *
 * `pagerStates` (preferred) renders a horizontal pager of canonical states.
 * `a11yReport` (preferred) opens A11ySheetContent on the A11y pill. The
 * legacy `states` and `a11y` slots remain for incomplete migrations.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
public fun PlaygroundScaffold(
    preview: @Composable () -> Unit,
    knobs: (@Composable () -> Unit)? = null,
    states: (@Composable () -> Unit)? = null,
    code: (() -> String)? = null,
    a11y: (@Composable () -> Unit)? = null,
    spec: (@Composable () -> Unit)? = null,
    pagerStates: List<PlaygroundState>? = null,
    a11yReport: A11yReport? = null,
    modifier: Modifier = Modifier,
) {
    var knobsOpen by rememberSaveable { mutableStateOf(false) }
    var a11yOpen by rememberSaveable { mutableStateOf(false) }
    var codeOpen by rememberSaveable { mutableStateOf(false) }

    val hasA11y = a11yReport != null || a11y != null

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp5),
    ) {
        PreviewSurface(modifier = Modifier.heightIn(min = 200.dp)) { preview() }

        if (knobs != null || hasA11y || code != null) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp3),
            ) {
                if (knobs != null) {
                    ActionPill(
                        iconRes = PrismaIcons.Edit,
                        label = "Edit",
                        modifier = Modifier.weight(1f),
                        onClick = { knobsOpen = true },
                    )
                }
                if (hasA11y) {
                    ActionPill(
                        iconRes = PrismaIcons.Eye,
                        label = "A11y",
                        modifier = Modifier.weight(1f),
                        onClick = { a11yOpen = true },
                    )
                }
                if (code != null) {
                    ActionPill(
                        iconRes = PrismaIcons.Doc,
                        label = "Code",
                        modifier = Modifier.weight(1f),
                        onClick = { codeOpen = true },
                    )
                }
            }
        }

        // Pager wins over legacy gallery — the swipeable layout is the new
        // default per the playground rollout brief.
        when {
            pagerStates != null && pagerStates.isNotEmpty() -> {
                PlaygroundSection(label = "States — swipe to compare") {
                    StatesPager(states = pagerStates)
                }
            }
            states != null -> {
                PlaygroundSection(label = "States") {
                    StatesGallery { states() }
                }
            }
        }

        if (spec != null) {
            PlaygroundSection(label = "Spec") {
                spec()
            }
        }
    }

    if (knobsOpen && knobs != null) {
        ModalBottomSheet(
            onDismissRequest = { knobsOpen = false },
            sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false),
            containerColor = PrismaSemanticColors.SurfaceBase.themed(),
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = PrismaSpacing.Sp5, vertical = PrismaSpacing.Sp4),
                verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp4),
            ) {
                Text(
                    text = "Edit",
                    style = PrismaTypography.HeadlineSm,
                    color = PrismaSemanticColors.TextPrimary.themed(),
                )
                knobs()
                PrismaButton(
                    text = "Done",
                    onClick = { knobsOpen = false },
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }

    if (a11yOpen && hasA11y) {
        ModalBottomSheet(
            onDismissRequest = { a11yOpen = false },
            sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false),
            containerColor = PrismaSemanticColors.SurfaceBase.themed(),
        ) {
            if (a11yReport != null) {
                A11ySheetContent(report = a11yReport)
            } else if (a11y != null) {
                Column(modifier = Modifier.fillMaxWidth()) { a11y() }
            }
        }
    }

    if (codeOpen && code != null) {
        ModalBottomSheet(
            onDismissRequest = { codeOpen = false },
            sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false),
            containerColor = PrismaSemanticColors.SurfaceBase.themed(),
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = PrismaSpacing.Sp5, vertical = PrismaSpacing.Sp4),
                verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp4),
            ) {
                Text(
                    text = "Usage",
                    style = PrismaTypography.HeadlineSm,
                    color = PrismaSemanticColors.TextPrimary.themed(),
                )
                Text(
                    text = "Drop this in to render the component as it appears above. Only knobs that differ from defaults are shown.",
                    style = PrismaTypography.BodySm,
                    color = PrismaSemanticColors.TextSecondary.themed(),
                )
                CodeBlock(code = code())
            }
        }
    }
}

@Composable
private fun PlaygroundSection(label: String, content: @Composable () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp4)) {
        Text(
            text = label.uppercase(),
            style = PrismaTypography.LabelSm,
            color = PrismaSemanticColors.TextTertiary.themed(),
        )
        content()
    }
}

@Composable
private fun ActionPill(
    iconRes: Int,
    label: String,
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
) {
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(PrismaRadius.Md))
            .background(PrismaSemanticColors.SurfaceRaised.themed())
            .border(1.dp, PrismaSemanticColors.BorderSubtle.themed(), RoundedCornerShape(PrismaRadius.Md))
            .clickable(role = Role.Button, onClick = onClick)
            .padding(horizontal = PrismaSpacing.Sp4, vertical = PrismaSpacing.Sp3),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2),
    ) {
        Icon(
            painter = painterResource(iconRes),
            contentDescription = null,
            tint = PrismaSemanticColors.TextSecondary.themed(),
            modifier = Modifier.size(16.dp),
        )
        Text(
            text = label,
            style = PrismaTypography.LabelMd,
            color = PrismaSemanticColors.TextPrimary.themed(),
        )
    }
}

File: ios-catalogue/App/Playground/PlaygroundScaffold.swift
import SwiftUI
import CoreUI
import Components

/// Storybook-style scaffold for an interactive component showcase.
///
/// Pre-existing showcases call this with a `knobs` block, an optional
/// `states` block (rendered as a flow gallery), a `code` generator, and
/// an optional `a11y` block. Knobs and a11y are now both surfaced behind
/// bottom sheets — opened by Edit / A11y action pills directly under the
/// preview — so the live preview never scrolls off screen behind a long
/// edit panel.
///
/// The richer ButtonShowcase uses [PlaygroundScreen] (with a structured
/// [A11yReport] and a horizontal states pager) directly, bypassing this
/// scaffold. Both APIs coexist during the per-component rollout.
struct PlaygroundScaffold<Preview: View, Knobs: View, States: View, A11y: View>: View {
    @Environment(\.colorScheme) private var scheme
    private let preview: () -> Preview
    private let knobs: () -> Knobs
    private let states: () -> States
    private let code: (() -> String)?
    private let a11y: () -> A11y

    init(
        @ViewBuilder preview: @escaping () -> Preview,
        @ViewBuilder knobs: @escaping () -> Knobs = { EmptyView() },
        @ViewBuilder states: @escaping () -> States = { EmptyView() },
        code: (() -> String)? = nil,
        @ViewBuilder a11y: @escaping () -> A11y = { EmptyView() }
    ) {
        self.preview = preview
        self.knobs = knobs
        self.states = states
        self.code = code
        self.a11y = a11y
    }

    @State private var knobsOpen = false
    @State private var a11yOpen = false

    private var hasKnobs: Bool { Knobs.self != EmptyView.self }
    private var hasA11y: Bool { A11y.self != EmptyView.self }

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp5) {
            PreviewSurface { preview() }
                .frame(minHeight: 200)

            if hasKnobs || hasA11y {
                HStack(spacing: PrismaSpacing.sp3) {
                    if hasKnobs {
                        actionPill(icon: .edit, label: "Edit") { knobsOpen = true }
                    }
                    if hasA11y {
                        actionPill(icon: .eye, label: "A11y") { a11yOpen = true }
                    }
                }
            }

            if States.self != EmptyView.self {
                section(label: "States") {
                    StatesGallery { states() }
                }
            }

            if let code {
                section(label: "Usage") {
                    CodeBlock(code: code())
                }
            }
        }
        .sheet(isPresented: $knobsOpen) {
            ScrollView {
                VStack(alignment: .leading, spacing: PrismaSpacing.sp4) {
                    Text("Edit")
                        .font(PrismaTypography.headlineSm.font)
                        .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                    knobs()
                    PrismaButton("Done") { knobsOpen = false }
                        .frame(maxWidth: .infinity)
                }
                .padding(.horizontal, PrismaSpacing.sp5)
                .padding(.vertical, PrismaSpacing.sp4)
            }
            .presentationDetents([.medium, .large])
            .presentationDragIndicator(.visible)
        }
        .sheet(isPresented: $a11yOpen) {
            ScrollView {
                VStack(alignment: .leading, spacing: PrismaSpacing.sp4) {
                    a11y()
                }
                .padding(.horizontal, PrismaSpacing.sp5)
                .padding(.vertical, PrismaSpacing.sp4)
            }
            .presentationDetents([.large])
            .presentationDragIndicator(.visible)
        }
    }

    @ViewBuilder
    private func section<Body: View>(label: String, @ViewBuilder content: () -> Body) -> some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp4) {
            Text(label.uppercased())
                .font(PrismaTypography.labelSm.font)
                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
            content()
        }
    }

    @ViewBuilder
    private func actionPill(icon: PrismaIcon, label: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: PrismaSpacing.sp2) {
                Image(prisma: icon)
                    .renderingMode(.template)
                    .resizable()
                    .frame(width: 16, height: 16)
                    .foregroundStyle(PrismaSemanticColors.textSecondary.themed(scheme))
                Text(label)
                    .font(PrismaTypography.labelMd.font)
                    .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
            }
            .frame(maxWidth: .infinity)
            .padding(.horizontal, PrismaSpacing.sp4)
            .padding(.vertical, PrismaSpacing.sp3)
            .background(PrismaSemanticColors.surfaceRaised.themed(scheme))
            .overlay(
                RoundedRectangle(cornerRadius: PrismaRadius.md)
                    .strokeBorder(PrismaSemanticColors.borderSubtle.themed(scheme), lineWidth: 1)
            )
            .clipShape(RoundedRectangle(cornerRadius: PrismaRadius.md))
        }
        .buttonStyle(.plain)
    }
}
